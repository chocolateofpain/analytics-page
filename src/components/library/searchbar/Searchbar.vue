<template>
  <div 
    class="searchbar"
    :style="{ width: this.focused ? (this.widthInput + 50) + 'px' : this.widthInput + 'px'}"

  >
    <Icon name="search" />
    <input
      @focus="focused = true"
      @blur="focused = false"
      @input="onInput"
      class="searchbar-input"
      name="searchbar"
      :placeholder="placeholder"
      ref="input"
      type="text"
      v-model="searchTerm"
    />
    <div
      @mousedown="onClickClearInput"
      class="clear-input"
      v-show="searchTerm.length > 0"
    >
      <Icon
        name="x"
        :width="18"
      />
    </div>
  </div>
</template>

<script>
import Icon from '../icon/Icon.vue'
export default {
  components: {
    Icon
  },
  computed: {
    searchTerm: {
      get() {
        return this.value;
      },
      set(value) {
        this.$emit('input', value);
      }
    }
  },
  data () {
    return {
      focused: false,
      showClearButton: false,
      widthInput: this.minWidth
    }
  },
  name: 'Searchbar',
  methods: {
    onClickClearInput () {
      this.searchTerm = ''
    },
    onInput () {
      if (this.searchTerm.length > 0) {
        this.showClearButton = true
      } else {
        this.showClearButton = false
      }
    },
    // TODO add UI hint for keyboard shortcut
    onSlashFocusInput: function (event) {
      if (event.key === '/') {
        this.$refs.input.focus();
      }
    }
  },
  mounted() {
    document.addEventListener('keyup', this.onSlashFocusInput);
  },
  beforeDestroy() {
    document.removeEventListener('keyup', this.onSlashFocusInput);
  },
  props: {
    /**
    * Adds a placeholder text
    */
    placeholder: {
      type: String,
      default: null
    },
    /**
    * Sets width of the Searchbar, 50px are added on focus
    */
    minWidth: {
      type: Number,
      default: 200
    },
    /**
    * Binds value of input in Searchbar component to v-model directive set on <Searchbar /> in parent component.
    * Example: \<Searchbar v-model="myVeryOwnVariableName" />
    */
    value: {
      type: String,
      required: true,
    }
  }
}
</script>

<style lang="scss" scoped>
// TODO: improve styling from user agent stylesheet for internal autofill
.searchbar {
  align-items: center;
  background: rgba(237, 237, 237, 1);
  // avoid jumpy UI on focus
  border: 1px solid rgba(237, 237, 237, 1);
  border-radius: 4px;
  box-sizing: border-box;
  color: #000;
  display: flex;
  height: 28px;
  margin: 4px;
  padding: 0 7px;
  position: relative;
}
.searchbar:focus-within {
  background-color: rgba(237, 237, 237, 0.4);
  border: 1px solid rgba(237, 237, 237, 1);
}
.searchbar-input {
  background: none;
  border: none;
  height: 100%;
  padding: 0 30px 0 30px;
  position: absolute;
  // removing this is not recommended and means implementing a visual cue for focus state
  // I haven't done this in this case since it's not part of the design
  outline: none;
  left: 0;
  width: 100%;
}
.clear-input {
  position: absolute;
  right: 0;
  width: 25px;
}
::placeholder { 
  color: #000;
  opacity: 1;
}
::-ms-input-placeholder {
  color: #000;
}
</style>
