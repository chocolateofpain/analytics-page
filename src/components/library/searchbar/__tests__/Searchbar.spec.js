import { mount } from '@vue/test-utils'
import Searchbar from '../Searchbar.vue'

describe('Searchbar', () => {
  it('should be defined', () => {
    expect(Searchbar).toBeDefined()
  })

  describe('snapshots', () => {
    it('should render default component', () => {
      const component = _mount({})
      expect(component).toMatchSnapshot()
    })

    it('should render with a custom width', () => {
      const component = _mount({ minWidth: 300 })
      expect(component).toMatchSnapshot()
    })

    it('should render with a custom placeholder', () => {
      const component = _mount({ placeholder: 'I\'m a random placeholder' })
      expect(component).toMatchSnapshot()
    })
  })

  describe('input functionality', () => {
    it('should update input', async () => {
      const component = _mount({})
      const searchbarInput = component.find('input[type="text"]')
      await searchbarInput.setValue('hello darkness my old friend')
      expect(component.find('input[type="text"]').element.value).toBe('hello darkness my old friend')
    })

    it('should call onClickClearInput', async () => {
      // TODO: improve test to not throw deprecated warning
      const onClickClearInputMethod = jest.spyOn(Searchbar.methods, 'onClickClearInput')
      const component = mount(Searchbar, {
        attrs: {
          value: ''
        },
        methods: onClickClearInputMethod
      })
      const searchbarInput = component.find('input[type="text"]')
      await searchbarInput.setValue('hello darkness my old friend')
      const clearInputButton = component.find('.clear-input')
      clearInputButton.trigger('mousedown', { left: true })
      expect(onClickClearInputMethod).toHaveBeenCalled()
    })
  })

  /**
   * Simple wrapper to avoid duplicated code
   */
  const _mount = (props) => {
    return mount(Searchbar, {
      propsData: props,
      attrs: {
        value: ''
      }
    })
  }
})
