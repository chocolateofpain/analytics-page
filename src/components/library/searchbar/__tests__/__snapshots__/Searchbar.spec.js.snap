// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar snapshots should render default component 1`] = `
<div class="searchbar" style="width: 200px;">
  <div class="icon"><svg version="1.1" width="16" height="16" viewBox="0 0 16 16" class="octicon octicon-search" aria-hidden="true">
      <path fill-rule="evenodd" d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"></path>
    </svg></div> <input name="searchbar" type="text" class="searchbar-input">
  <div class="clear-input" style="display: none;">
    <div class="icon"><svg version="1.1" width="18" height="18" viewBox="0 0 16 16" class="octicon octicon-x" aria-hidden="true">
        <path fill-rule="evenodd" d="M3.72 3.72a.75.75 0 011.06 0L8 6.94l3.22-3.22a.75.75 0 111.06 1.06L9.06 8l3.22 3.22a.75.75 0 11-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 01-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 010-1.06z"></path>
      </svg></div>
  </div>
</div>
`;

exports[`Searchbar snapshots should render with a custom placeholder 1`] = `
<div class="searchbar" style="width: 200px;">
  <div class="icon"><svg version="1.1" width="16" height="16" viewBox="0 0 16 16" class="octicon octicon-search" aria-hidden="true">
      <path fill-rule="evenodd" d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"></path>
    </svg></div> <input name="searchbar" placeholder="I'm a random placeholder" type="text" class="searchbar-input">
  <div class="clear-input" style="display: none;">
    <div class="icon"><svg version="1.1" width="18" height="18" viewBox="0 0 16 16" class="octicon octicon-x" aria-hidden="true">
        <path fill-rule="evenodd" d="M3.72 3.72a.75.75 0 011.06 0L8 6.94l3.22-3.22a.75.75 0 111.06 1.06L9.06 8l3.22 3.22a.75.75 0 11-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 01-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 010-1.06z"></path>
      </svg></div>
  </div>
</div>
`;

exports[`Searchbar snapshots should render with a custom width 1`] = `
<div class="searchbar" style="width: 300px;">
  <div class="icon"><svg version="1.1" width="16" height="16" viewBox="0 0 16 16" class="octicon octicon-search" aria-hidden="true">
      <path fill-rule="evenodd" d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"></path>
    </svg></div> <input name="searchbar" type="text" class="searchbar-input">
  <div class="clear-input" style="display: none;">
    <div class="icon"><svg version="1.1" width="18" height="18" viewBox="0 0 16 16" class="octicon octicon-x" aria-hidden="true">
        <path fill-rule="evenodd" d="M3.72 3.72a.75.75 0 011.06 0L8 6.94l3.22-3.22a.75.75 0 111.06 1.06L9.06 8l3.22 3.22a.75.75 0 11-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 01-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 010-1.06z"></path>
      </svg></div>
  </div>
</div>
`;
